### PR TITLE
Implement LocaleAwareInterface to remove RequestStack dependency

### LINKED ISSUE #108: The translation extractor fail with DatePickerType in form
#### Sonata packages

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle              3.72.0 3.72.0 The missing Symfony Admin Generator
sonata-project/block-bundle              3.20.0 4.2.0  Symfony SonataBlockBundle
sonata-project/cache                     2.0.1  2.0.1  Cache library
sonata-project/doctrine-extensions       1.8.0  1.8.0  Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.21.0 3.21.0 Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/exporter                  2.3.0  2.3.0  Lightweight Exporter library
sonata-project/form-extensions           1.5.0  1.5.0  Symfony form extensions
sonata-project/intl-bundle               2.7.0  2.7.0  Symfony SonataIntlBundle
sonata-project/twig-extensions           1.3.1  1.3.1  Sonata twig extensions
```

#### Symfony packages

```
$ composer show --latest 'symfony/*'
symfony/asset                      v4.4.11 v4.4.11 Symfony Asset Component
symfony/browser-kit                v4.4.11 v4.4.11 Symfony BrowserKit Component
symfony/cache                      v4.4.11 v4.4.11 Symfony Cache component with PSR-6, PSR-16, and tags
symfony/cache-contracts            v2.1.3  v2.1.3  Generic abstractions related to caching
symfony/config                     v4.4.11 v4.4.11 Symfony Config Component
symfony/console                    v4.4.11 v4.4.11 Symfony Console Component
symfony/css-selector               v4.4.11 v4.4.11 Symfony CssSelector Component
symfony/debug                      v4.4.11 v4.4.11 Symfony Debug Component
symfony/debug-bundle               v4.4.11 v4.4.11 Symfony DebugBundle
symfony/debug-pack                 v1.0.8  v1.0.8  A debug pack for Symfony projects
symfony/dependency-injection       v4.4.11 v4.4.11 Symfony DependencyInjection Component
symfony/doctrine-bridge            v4.4.11 v4.4.11 Symfony Doctrine Bridge
symfony/dom-crawler                v4.4.11 v4.4.11 Symfony DomCrawler Component
symfony/dotenv                     v4.4.11 v4.4.11 Registers environment variables from a .env file
symfony/error-handler              v4.4.11 v4.4.11 Symfony ErrorHandler Component
symfony/event-dispatcher           v4.4.11 v4.4.11 Symfony EventDispatcher Component
symfony/event-dispatcher-contracts v1.1.9  v2.1.3  Generic abstractions related to dispatching event
symfony/expression-language        v4.4.11 v4.4.11 Symfony ExpressionLanguage Component
symfony/filesystem                 v4.4.11 v4.4.11 Symfony Filesystem Component
symfony/finder                     v4.4.11 v4.4.11 Symfony Finder Component
symfony/flex                       v1.9.1  v1.9.1  Composer plugin for Symfony
symfony/form                       v4.4.11 v4.4.11 Symfony Form Component
symfony/framework-bundle           v4.4.11 v4.4.11 Symfony FrameworkBundle
symfony/http-client                v4.4.11 v4.4.11 Symfony HttpClient component
symfony/http-client-contracts      v2.1.3  v2.1.3  Generic abstractions related to HTTP clients
symfony/http-foundation            v4.4.11 v4.4.11 Symfony HttpFoundation Component
symfony/http-kernel                v4.4.11 v4.4.11 Symfony HttpKernel Component
symfony/inflector                  v4.4.11 v4.4.11 Symfony Inflector Component
symfony/intl                       v4.4.11 v4.4.11 A PHP replacement layer for the C intl extension that includes additional data from the ICU library.
symfony/maker-bundle               v1.20.0 v1.20.0 Symfony Maker helps you create empty commands, controllers, form classes, tests and more so you can forget about writing boilerplate code.
symfony/mime                       v4.4.11 v4.4.11 A library to manipulate MIME messages
symfony/monolog-bridge             v4.4.11 v4.4.11 Symfony Monolog Bridge
symfony/monolog-bundle             v3.5.0  v3.5.0  Symfony MonologBundle
symfony/options-resolver           v4.4.11 v4.4.11 Symfony OptionsResolver Component
symfony/orm-pack                   v2.0.0  v2.0.0  A pack for the Doctrine ORM
symfony/phpunit-bridge             v5.1.3  v5.1.3  Symfony PHPUnit Bridge
symfony/polyfill-intl-grapheme     v1.18.0 v1.18.0 Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu          v1.18.0 v1.18.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn          v1.18.0 v1.18.0 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer   v1.18.0 v1.18.0 Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring          v1.18.0 v1.18.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php72             v1.18.0 v1.18.0 Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-php73             v1.18.0 v1.18.0 Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/polyfill-php80             v1.18.0 v1.18.0 Symfony polyfill backporting some PHP 8.0+ features to lower PHP versions
symfony/process                    v4.4.11 v4.4.11 Symfony Process Component
symfony/profiler-pack              v1.0.4  v1.0.4  A pack for the Symfony web profiler
symfony/property-access            v4.4.11 v4.4.11 Symfony PropertyAccess Component
symfony/property-info              v4.4.11 v4.4.11 Symfony Property Info Component
symfony/routing                    v4.4.11 v4.4.11 Symfony Routing Component
symfony/security-acl               v3.0.4  v3.0.4  Symfony Security Component - ACL (Access Control List)
symfony/security-bundle            v4.4.11 v4.4.11 Symfony SecurityBundle
symfony/security-core              v4.4.11 v4.4.11 Symfony Security Component - Core Library
symfony/security-csrf              v4.4.11 v4.4.11 Symfony Security Component - CSRF Library
symfony/security-guard             v4.4.11 v4.4.11 Symfony Security Component - Guard
symfony/security-http              v4.4.11 v4.4.11 Symfony Security Component - HTTP Integration
symfony/serializer                 v4.4.11 v4.4.11 Symfony Serializer Component
symfony/serializer-pack            v1.0.3  v1.0.3  A pack for the Symfony serializer
symfony/service-contracts          v2.1.3  v2.1.3  Generic abstractions related to writing services
symfony/stopwatch                  v4.4.11 v4.4.11 Symfony Stopwatch Component
symfony/string                     v5.1.3  v5.1.3  Symfony String component
symfony/swiftmailer-bundle         v3.4.0  v3.4.0  Symfony SwiftmailerBundle
symfony/templating                 v4.4.11 v4.4.11 Symfony Templating Component
symfony/test-pack                  v1.0.6  v1.0.6  A pack for functional and end-to-end testing within a Symfony app
symfony/translation                v4.4.11 v4.4.11 Symfony Translation Component
symfony/translation-contracts      v2.1.3  v2.1.3  Generic abstractions related to translation
symfony/twig-bridge                v4.4.11 v4.4.11 Symfony Twig Bridge
symfony/twig-bundle                v4.4.11 v4.4.11 Symfony TwigBundle
symfony/validator                  v4.4.11 v4.4.11 Symfony Validator Component
symfony/var-dumper                 v4.4.11 v4.4.11 Symfony mechanism for exploring and dumping PHP variables
symfony/var-exporter               v4.4.11 v4.4.11 A blend of var_export() + serialize() to turn any serializable data structure to plain PHP code
symfony/web-link                   v4.4.11 v4.4.11 Symfony WebLink Component
symfony/web-profiler-bundle        v4.4.11 v4.4.11 Symfony WebProfilerBundle
symfony/webpack-encore-bundle      v1.7.3  v1.7.3  Integration with your Symfony app & Webpack Encore!
symfony/yaml                       v4.4.11 v4.4.11 Symfony Yaml Component
```

#### PHP version

```
$ php -v
PHP 7.4.8 (cli) (built: Jul  9 2020 23:21:18) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.8, Copyright (c), by Zend Technologies
```

## Subject

Hi, after a dependency upgrade i got an Exception "[LogicException] A Request must be available." when using the `translation:update` command.

It comes from using the type "DatePickerType" in one of my forms, as soon as I remove it, no more error.
It seems the parent "BasePickerType" tries to access the locale from the "StackRequest" object in parameter of the constructor but in the context of a console command it's always null ?

Sorry if I missed something, thank you for your work.

## Steps to reproduce

In the configureFormFields() method of an admin, just add a field with "DatePickerType" as type and launch the translation:update command:

```
protected function configureFormFields(FormMapper $formMapper): void
{
        $formMapper->add('activityMonth', DatePickerType::class, [
                'label' => 'activity_month_MM_YYYY',
                'format' => 'MM/yyyy',
         ]);
}
```

## Actual results

```
$ php bin/console translation:update en --dump-messages -vvv

In BasePickerType.php line 160:
                                
  [LogicException]              
  A Request must be available.  
                                

Exception trace:
  at /var/www/vendor/sonata-project/form-extensions/src/Type/BasePickerType.php:160
 Sonata\Form\Type\BasePickerType->getLocale() at /var/www/vendor/sonata-project/form-extensions/src/Type/BasePickerType.php:53
 Sonata\Form\Type\BasePickerType->__construct() at /var/www/var/cache/dev/ContainerF9sHqCT/srcApp_KernelDevDebugContainer.php:11349
 ContainerF9sHqCT\srcApp_KernelDevDebugContainer->getSonata_Form_Type_DatePickerService() at /var/www/vendor/symfony/dependency-injection/Container.php:450
 Symfony\Component\DependencyInjection\Container->getService() at /var/www/vendor/symfony/dependency-injection/Argument/ServiceLocator.php:40
 Symfony\Component\DependencyInjection\Argument\ServiceLocator->get() at /var/www/vendor/symfony/form/Extension/DependencyInjection/DependencyInjectionExtension.php:43
 Symfony\Component\Form\Extension\DependencyInjection\DependencyInjectionExtension->getType() at /var/www/vendor/symfony/form/FormRegistry.php:77
 Symfony\Component\Form\FormRegistry->getType() at /var/www/vendor/symfony/form/FormFactory.php:74
 Symfony\Component\Form\FormFactory->createNamedBuilder() at /var/www/vendor/symfony/form/FormBuilder.php:94
 Symfony\Component\Form\FormBuilder->create() at /var/www/vendor/symfony/form/FormBuilder.php:244
 Symfony\Component\Form\FormBuilder->resolveChildren() at /var/www/vendor/symfony/form/FormBuilder.php:195
 Symfony\Component\Form\FormBuilder->getForm() at /var/www/vendor/sonata-project/admin-bundle/src/Admin/AbstractAdmin.php:3394
 Sonata\AdminBundle\Admin\AbstractAdmin->buildForm() at /var/www/vendor/sonata-project/admin-bundle/src/Admin/AbstractAdmin.php:1463
 Sonata\AdminBundle\Admin\AbstractAdmin->getForm() at /var/www/vendor/sonata-project/admin-bundle/src/Translator/Extractor/AdminExtractor.php:101
 Sonata\AdminBundle\Translator\Extractor\AdminExtractor->extract() at /var/www/vendor/symfony/translation/Extractor/ChainExtractor.php:56
 Symfony\Component\Translation\Extractor\ChainExtractor->extract() at /var/www/vendor/symfony/framework-bundle/Command/TranslationUpdateCommand.php:226
 Symfony\Bundle\FrameworkBundle\Command\TranslationUpdateCommand->execute() at /var/www/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /var/www/vendor/symfony/console/Application.php:1018
 Symfony\Component\Console\Application->doRunCommand() at /var/www/vendor/symfony/framework-bundle/Console/Application.php:97
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /var/www/vendor/symfony/console/Application.php:271
 Symfony\Component\Console\Application->doRun() at /var/www/vendor/symfony/framework-bundle/Console/Application.php:83
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/vendor/symfony/console/Application.php:147
 Symfony\Component\Console\Application->run() at /var/www/bin/console:42

translation:update [--prefix [PREFIX]] [--output-format [OUTPUT-FORMAT]] [--dump-messages] [--force] [--no-backup] [--clean] [--domain [DOMAIN]] [--xliff-version [XLIFF-VERSION]] [--sort [SORT]] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command> <locale> [<bundle>]
```

### PR DESCRIPTION
**Pull request by franmomu**
## Subject

`BasePickerType` needed a `RequestStack` object and it didn't work without a Request. 

With this change, `BasePickerType` receives a default locale (`kernel.default_locale` parameter) and when the request is made, the [LocalAwareListener](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php) (through `LocalAwareInterface` and `kernel.locale_aware` tag) set the locale from the Request.

I am targeting this branch, because these changes are BC.

Closes #108.

## Changelog

```markdown
### Deprecated
- Deprecated passing a `RequestStack` object to `BasePickerType` as third parameter, the default locale should be passed instead.
### Fixed
- Fixed using `BasePickerType` without a request.
```

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Should we make this PR on the 0.x branch ?

**Comment by jordisala1991** [comment]
IMO we have a migration path for 1.x compatible with all bundles. No one should use the 0.x anymore. I would say to go for 1.x (different thing with block bundle, where we have 4.x that nobody uses because none of our bundles are compatible)

File: UPGRADE-1.x.md [hunk 1]
```diff
@@ -1,6 +1,11 @@
 UPGRADE 1.x
 ===========
 
+## Sonata\Form\Type\BasePickerType
+
+Deprecate passing a `RequestStack` object as third parameter, you MUST pass a default locale instead.
+
+
```
Comment (phansys) [inline UPGRADE-1.x.md:8]:
```suggestion
```

File: src/Type/BasePickerType.php [hunk 2]
```diff
@@ -45,12 +46,34 @@ abstract class BasePickerType extends AbstractType
...
     {
         $this->formatConverter = $formatConverter;
         $this->translator = $translator;
 
-        $this->locale = $this->getLocale($requestStack);
+        // NEXT_MAJOR: Remove this block
+        if (!\is_string($requestStackOrDefaultLocale) && !$requestStackOrDefaultLocale instanceof RequestStack) {
+            throw new \InvalidArgumentException(sprintf(
+                'Argument 3 should be a "string" or instance of "%s"',
+                RequestStack::class
```
Comment (phansys) [inline src/Type/BasePickerType.php:61]:
```suggestion
                'Argument 3 passed to "%s()" be must be of type string or an instance of %s, %s given.',
                __METHOD__,
                \is_object($requestStackOrDefaultLocale) ? 'instance of '.\get_class($requestStackOrDefaultLocale) : \gettype($requestStackOrDefaultLocale),
                RequestStack::class
```

**Comment by franmomu** [comment]
> Should we make this PR on the 0.x branch ?

[BasePickerType from 0.x branch](https://github.com/sonata-project/form-extensions/blob/0.x/src/Type/BasePickerType.php)  is different from 1.x, it doesn't have the `RequestStack` parameter.

> Cause 1.x is currently not used by SonataAdmin.

[hmm what do you mean?](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/composer.json#L31)

```
"sonata-project/form-extensions": "^0.1.1 || ^1.4",
```

**Comment by VincentLanglet** [comment]
> IMO we have a migration path for 1.x compatible with all bundles. No one should use the 0.x anymore. I would say to go for 1.x (different thing with block bundle, where we have 4.x that nobody uses because none of our bundles are compatible)

> > Cause 1.x is currently not used by SonataAdmin.
> 
> [hmm what do you mean?](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/composer.json#L31)
> 
> ```
> "sonata-project/form-extensions": "^0.1.1 || ^1.4",
> ```

Oh ! I didn't know. Seems like I should update `form-extensions` on my project 😄

**Comment by greg0ire** [comment]
Thanks @franmomu !

**Pull request merged**